### PR TITLE
Template system refactoring

### PR DESCRIPTION
**Pull request by carlosescri**
See comment at the end of the commit list.

### CODE REVIEW

**Comment by carlosescri** [comment]
I think I finished with the new templating system. I'm going to investigate the last two points in this PR. The last one shouldn't take too much time. The XML entities stuff involves regular expressions and may take some time.

**Comment by carlosescri** [comment]
Well, I have to add docs too because after a month I won't know all the details about this.

**Comment by carlosescri** [comment]
**TLDR;**

- No templates in slots.
- No forced patterns for templates (the old `Widget.renderEmpty()`, `Widget.renderNoContent()`, `Widget.renderContent() methods`). As a component author you can follow recommended patterns but they're not enforced in JS.
- Templates are declared with a `@template` decorator, with the same parameters as `@property` but some of them are forced internally.
- Templates from JS perspective are functions with this signature: <br>`(html: Function, data: GenericObject): TemplateResult => { return html…; }`
- `this` is available inside the functions and correspond to the widget rendering the template.
- Templates can be assigned via attributes by passing a unique CSS id (`#myTemplate`), or via JS properties by passing a unique CSS id, an _unbound_ function with the right signature (bound functions will cause problems), or a string representing a valid template.
- Templates declared in HTML use the `<template>` tag.

**Recommended patterns for declaring templates**

```typescript
// main template
@template({ attribute: 'template' }) 
public renderTemplate: TemplateFunction = DEFAULT_TEMPLATE;

// secondary template
@template({ attribute: 'no-content-template' })
public renderNoContentTemplate: TemplateFunction = DEFAULT_NO_CONTENT_TEMPLATE;

// template for a child component
@property({ type: String, attribute: 'card-template-id' })
public cardTemplateId: CSSSelector = null;
```

**Template Tag Sample**

```html
<template id="displayTemplate">
  ${
    data.total > 0
    ? html`<p>${data.total} result${data.total > 1 ? 's' : ''} found (&amp; &lt; &gt;)!</p>`
    : html`<p>Nothing to display…</p>`
  }

  <!--
    "this" can't be used if no explicit handle[WhateverEvent] method has been
    declared in the component!!!
  -->
  <button type="button" @click=${this}>Come on and break the app!</button>
</template>
```

**Comment by resetreboot** [comment]
Good refactor. Looks way simpler than the previous templating and rendering functions indeed.
